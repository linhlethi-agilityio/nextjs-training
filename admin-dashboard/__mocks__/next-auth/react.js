export const useSession = () => ({
  data: {
    user: {
      token: 'access-token',
      email: 'user@gmail.com',
      id: '1',
      name: 'User',
      role: 'admin',
    },
    expires: Date.now().toString(),
  },
  status: 'authenticated',
  update: async () => null,
});
