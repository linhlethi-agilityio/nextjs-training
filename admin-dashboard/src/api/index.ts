export * from './orders';
export * from './customers';
