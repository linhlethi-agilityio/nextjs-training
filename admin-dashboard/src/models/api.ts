export type ResponseData<T> = {
  data: T;
};
