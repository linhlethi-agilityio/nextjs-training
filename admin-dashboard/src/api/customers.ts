import { API_ENDPOINT, DEFAULT_LIMIT } from '@/constants';
import { Customer } from '@/models';
import { api } from '@/services';

interface params {
  limit?: number;
  query?: string;
  page?: number;
}

export const getCustomers = async (params?: params) => {
  try {
    const { limit = DEFAULT_LIMIT, query = '', page = 1 } = params || {};

    const data = await api.getData<Customer[]>(
      API_ENDPOINT.CUSTOMERS,
      {
        limit,
        name: query,
        page,
      },
      {
        next: { tags: ['customers'], revalidate: 3600 },
      },
    );

    return {
      data: data.data ?? [],
    };
  } catch (error) {
    return { error };
  }
};

export const getTotalCustomers = async () => {
  try {
    const data = await api.getData<Customer[]>(
      API_ENDPOINT.CUSTOMERS,
      undefined,
      {
        next: { tags: ['customers'], revalidate: 3600 },
      },
    );

    return {
      data: data.data ?? [],
    };
  } catch (error) {
    return { error };
  }
};
