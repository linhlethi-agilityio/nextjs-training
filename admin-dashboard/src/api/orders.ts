// Services
import { API_ENDPOINT, DEFAULT_LIMIT, SORT_BY, SORT_ORDER } from '@/constants';

// Models
import { Order } from '@/models';

// Services
import { api } from '@/services';

interface params {
  limit?: number;
  query?: string;
  page?: number;
  sortBy?: SORT_BY;
  sortOrder?: SORT_ORDER;
}

export const getOrders = async (params?: params) => {
  try {
    const {
      limit = DEFAULT_LIMIT,
      query = '',
      page = 1,
      sortBy,
      sortOrder,
    } = params || {};

    const data = await api.getData<Order[]>(
      API_ENDPOINT.ORDERS,
      {
        limit,
        idOrder: query,
        page,
        ...(sortBy && sortOrder && { sortBy, order: sortOrder }),
      },
      {
        next: { tags: ['orders'] },
      },
    );

    return {
      data: data.data ?? [],
    };
  } catch (error) {
    return { error };
  }
};

export const getTotalOrders = async () => {
  try {
    const data = await api.getData<Order[]>(API_ENDPOINT.ORDERS, undefined, {
      next: { tags: ['orders'] },
    });

    return {
      data: data.data ?? [],
    };
  } catch (error) {
    return { error };
  }
};

export const getOrderById = async (id: string) => {
  try {
    const data = await api.getData<Order>(
      `${API_ENDPOINT.ORDERS}/${id}`,
      undefined,
      {
        next: { tags: ['order'] },
      },
    );

    return {
      data: data.data ?? [],
    };
  } catch (error) {
    return { error };
  }
};
