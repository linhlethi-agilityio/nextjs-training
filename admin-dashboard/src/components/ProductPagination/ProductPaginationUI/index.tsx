'use client';

import { memo, useCallback, useMemo } from 'react';
import { usePathname, useRouter, useSearchParams } from 'next/navigation';

// Components
import { Pagination } from '@/components';

interface ProductPaginationProps {
  totalPage: number;
}

const ProductPaginationUI = ({ totalPage }: ProductPaginationProps) => {
  const pathname = usePathname();
  const searchParams = useSearchParams();
  const { replace } = useRouter();

  const params = useMemo(
    () => new URLSearchParams(searchParams),
    [searchParams],
  );

  const currentPage = useMemo(
    () => (searchParams.get('page') ? Number(searchParams.get('page')) : 1),
    [searchParams],
  );

  const handleOnClickPagination = useCallback(
    (page: string) => {
      params.set('page', page);

      replace(`${pathname}?${params.toString()}`);
    },
    [params, pathname, replace],
  );

  return (
    <Pagination
      totalPage={totalPage}
      onClickPage={handleOnClickPagination}
      currentPage={currentPage}
    />
  );
};

export default memo(ProductPaginationUI);
