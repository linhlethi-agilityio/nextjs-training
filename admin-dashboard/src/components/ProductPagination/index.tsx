import { memo } from 'react';

// Components
import ProductPaginationUI from './ProductPaginationUI';

// Api
import { getTotalOrders } from '@/api';

// Constants
import { DEFAULT_LIMIT } from '@/constants';

interface ProductPaginationProps {
  limit: number;
}

const ProductPagination = async ({ limit }: ProductPaginationProps) => {
  const { data: totalOrders = [] } = await getTotalOrders();

  const totalPage = Math.ceil(totalOrders?.length / (limit ?? DEFAULT_LIMIT));

  return <ProductPaginationUI totalPage={totalPage} />;
};

export default memo(ProductPagination);
