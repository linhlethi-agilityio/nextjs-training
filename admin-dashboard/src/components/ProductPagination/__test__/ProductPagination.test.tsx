import { render } from '@testing-library/react';
import '@testing-library/jest-dom';

// Components
import ProductPagination from '..';

// Api
import * as api from '@/api';

jest.mock('@/api');

const mockReplace = jest.fn();
const mockURLSearchParams = {
  page: '1',
};

jest.mock('next/navigation', () => ({
  ...jest.requireActual('next/navigation'),
  usePathname: jest.fn(),
  useRouter: jest.fn(() => ({ replace: mockReplace })),
  useSearchParams: jest.fn(() => new URLSearchParams(mockURLSearchParams)),
}));

const props = {
  limit: 10,
};

describe('ProductPagination component', () => {
  it('should match snapshot for ProductPagination', async () => {
    jest.spyOn(api, 'getTotalOrders').mockResolvedValue({
      data: [
        {
          createdAt: '2024-07-23T00:33:27.809Z',
          idOrder: '969-875',
          product: 'fdgf',
          customer: 'Angie McCullough',
          status: 'Rejected',
          deadline: '2024-07-31',
          price: 3244,
          id: '1',
          customerId: '11',
          productImage:
            'https://images.pexels.com/photos/4041392/pexels-photo-4041392.jpeg?cs=srgb&dl=pexels-karolina-grabowska-4041392.jpg&fm=jpg',
        },
        {
          createdAt: '2024-07-23T00:33:46.006Z',
          idOrder: '529-297',
          product: 'aa',
          customer: 'Darrin Williamson',
          status: 'Pending',
          deadline: '2024-08-01',
          price: 32432,
          id: '2',
          customerId: '11',
          productImage:
            'https://images.pexels.com/photos/4041392/pexels-photo-4041392.jpeg?cs=srgb&dl=pexels-karolina-grabowska-4041392.jpg&fm=jpg',
        },
      ],
    });

    const ProductPaginationRender = await ProductPagination({ ...props });

    const { container } = render(ProductPaginationRender);

    expect(container).toMatchSnapshot();
  });
});
