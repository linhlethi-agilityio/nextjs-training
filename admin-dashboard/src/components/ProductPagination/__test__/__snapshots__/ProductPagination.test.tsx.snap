// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductPagination component should match snapshot for ProductPagination 1`] = `
<div>
  <div
    class="chakra-stack css-z424u0"
  >
    <button
      aria-label="pagination-prev"
      class="chakra-button css-4xx2wk"
      disabled=""
      type="button"
    >
      <svg
        class="chakra-icon css-nyt2fo"
        focusable="false"
        viewBox="0 0 8 14"
      >
        <path
          clip-rule="evenodd"
          d="M7.77789 0.979322C7.82445 1.02577 7.8614 1.08094 7.8866 1.14169C7.91181 1.20243 7.92478 1.26755 7.92478 1.33332C7.92478 1.39909 7.91181 1.46421 7.8866 1.52496C7.8614 1.5857 7.82445 1.64088 7.77789 1.68732L2.13089 7.33332L7.77789 12.9793C7.87178 13.0732 7.92452 13.2005 7.92452 13.3333C7.92452 13.4661 7.87178 13.5934 7.77789 13.6873C7.684 13.7812 7.55667 13.834 7.42389 13.834C7.29111 13.834 7.16378 13.7812 7.06989 13.6873L1.06989 7.68732C1.02333 7.64088 0.986384 7.5857 0.961177 7.52495C0.935971 7.46421 0.922997 7.39909 0.922997 7.33332C0.922997 7.26755 0.935971 7.20243 0.961177 7.14169C0.986384 7.08094 1.02333 7.02577 1.06989 6.97932L7.06989 0.979322C7.11634 0.932759 7.17151 0.895816 7.23226 0.870609C7.293 0.845403 7.35812 0.832428 7.42389 0.832428C7.48966 0.832428 7.55478 0.845403 7.61552 0.870609C7.67627 0.895816 7.73144 0.932759 7.77789 0.979322Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </button>
    <button
      class="chakra-button css-122nk53"
      type="button"
    >
      1
    </button>
    <button
      aria-label="pagination-next"
      class="chakra-button css-4xx2wk"
      disabled=""
      type="button"
    >
      <svg
        class="chakra-icon css-nyt2fo"
        focusable="false"
        viewBox="0 0 8 14"
      >
        <path
          clip-rule="evenodd"
          d="M0.797421 0.979322C0.843867 0.932759 0.899043 0.895816 0.959788 0.870609C1.02053 0.845403 1.08565 0.832428 1.15142 0.832428C1.21719 0.832428 1.28231 0.845403 1.34306 0.870609C1.4038 0.895816 1.45898 0.932759 1.50542 0.979322L7.50542 6.97932C7.55198 7.02577 7.58893 7.08094 7.61413 7.14169C7.63934 7.20243 7.65232 7.26755 7.65232 7.33332C7.65232 7.39909 7.63934 7.46421 7.61413 7.52495C7.58893 7.5857 7.55198 7.64088 7.50542 7.68732L1.50542 13.6873C1.41154 13.7812 1.2842 13.834 1.15142 13.834C1.01865 13.834 0.891308 13.7812 0.797421 13.6873C0.703535 13.5934 0.65079 13.4661 0.65079 13.3333C0.65079 13.2005 0.703535 13.0732 0.797421 12.9793L6.44442 7.33332L0.797421 1.68732C0.750858 1.64088 0.713915 1.5857 0.688709 1.52496C0.663502 1.46421 0.650528 1.39909 0.650528 1.33332C0.650528 1.26755 0.663502 1.20243 0.688709 1.14169C0.713915 1.08094 0.750858 1.02577 0.797421 0.979322Z"
          fill="currentColor"
          fill-rule="evenodd"
        />
      </svg>
    </button>
  </div>
</div>
`;
